cmd/kafka-consumer/writer.go

```diff
@@ -146,391 +101,382 @@ func newWriter(ctx context.Context, o *option) *writer {
+			zap.Any("protocol", w.protocol), zap.Bool("IsPartition", dml.TableInfo.TableName.IsPartition))
 		return
 	}
-	if row.CommitTs >= group.highWatermark {
-		group.Append(row, offset)
+	if commitTs >= group.highWatermark {
+		group.Append(dml)
+		log.Info("DML event append to the group",
+			zap.Int32("partition", group.partition),
+			zap.Any("offset", offset),
+			zap.Uint64("commitTs", dml.GetCommitTs()),
+			zap.Uint64("highWatermark", group.highWatermark),
+			zap.Int64("tableID", dml.GetTableID()),
+			zap.String("schema", dml.TableInfo.GetSchemaName()),
+			zap.String("table", dml.TableInfo.GetTableName()))
+		// zap.Any("columns", row.Columns), zap.Any("preColumns", row.PreColumns))
 		return
 	}
-	switch w.option.protocol {
+	switch w.protocol {
 	case config.ProtocolSimple, config.ProtocolOpen, config.ProtocolCanalJSON, config.ProtocolDebezium:
 		// simple protocol set the table id for all row message, it can be known which table the row message belongs to,
 		// also consider the table partition.
 		// open protocol set the partition table id if the table is partitioned.
 		// for normal table, the table id is generated by the fake table id generator by using schema and table name.
 		// so one event group for one normal table or one table partition, replayed messages can be ignored.
-		log.Warn("RowChangedEvent fallback row, since less than the group high watermark, ignore it",
-			zap.Int64("tableID", tableID), zap.Int32("partition", partition),
-			zap.Uint64("commitTs", row.CommitTs), zap.Any("offset", offset),
+		log.Warn("DML event fallback row, since less than the group high watermark, ignore it",
+			zap.Int64("tableID", tableID), zap.Int32("partition", progress.partition),
+			zap.Uint64("commitTs", commitTs), zap.Any("offset", offset),
 			zap.Uint64("highWatermark", group.highWatermark),
-			zap.Any("partitionWatermark", watermark), zap.Any("watermarkOffset", progress.watermarkOffset),
-			zap.String("schema", row.TableInfo.GetSchemaName()), zap.String("table", row.TableInfo.GetTableName()),
-			zap.Any("columns", row.Columns), zap.Any("preColumns", row.PreColumns),
-			zap.String("protocol", w.option.protocol.String()), zap.Bool("IsPartition", row.TableInfo.TableName.IsPartition))
+			zap.Any("partitionWatermark", progress.watermark), zap.Any("watermarkOffset", progress.watermarkOffset),
+			zap.String("schema", dml.TableInfo.GetSchemaName()), zap.String("table", dml.TableInfo.GetTableName()),
+			// zap.Any("columns", row.Columns), zap.Any("preColumns", row.PreColumns),
+			zap.Any("protocol", w.protocol), zap.Bool("IsPartition", dml.TableInfo.TableName.IsPartition))
 		return
 	default:
 	}
-	log.Warn("RowChangedEvent fallback row, since less than the group high watermark, do not ignore it",
-		zap.Int64("tableID", tableID), zap.Int32("partition", partition),
-		zap.Uint64("commitTs", row.CommitTs), zap.Any("offset", offset),
+	group.Append(dml)
```

Comment:
Why does the group append the old DML event?

Reply:
This can be removed after the partition table scenario is fully supported.